src/field/group.rs

```diff
@@ -0,0 +1,135 @@
+//! Groups $G$ are algebraic structures which are set and has a binary operation $\oplus$ that
+//! combines two elements $a, b$ of the set to produce a third element $a\oplus b$ in the set.
+//! The operation is said to have following properties:
+//! 1. Closure: $a\oplus b=c\in G$
+//! 2. Associative: $(a\oplus b)\oplus c = a\oplus(b\oplus c)$
+//! 3. Existence of Identity element: $a\oplus 0 = 0\oplus a = a$
+//! 4. Existence of inverse element for every element of the set: $a\oplus b=0$
+//! 5. Commutativity: Groups which satisfy an additional property: *commutativity* on the set of
+//!    elements are known as **Abelian groups**.
+//!
+//! [`FiniteGroup`] trait is implemented for all finite groups.
+use std::ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use super::{prime::find_primitive_element, FiniteField, PrimeField};
+
+#[const_trait]
+/// Group trait with finite number of elements
+pub trait FiniteGroup:
+  std::fmt::Debug
+  + Sized
+  + Add<Output = Self>
+  + AddAssign
+  + Sub<Output = Self>
+  + SubAssign
+  + Neg
+  + Mul<Self::Scalar, Output = Self>
+  + MulAssign<Self::Scalar>
+  + Clone
+  + Copy
+  + Default
+  + Eq {
+  /// Scalar defined in the group
+  type Scalar: FiniteField;
+  /// Identity element of group
+  const IDENTITY: Self;
+  /// total number of elements in group
+  const ORDER: usize;
+  /// primtive element of group
+  const GENERATOR: Self;
```

Comment:
Not every finite group has a generator!

Reply:
Added a new trait `FiniteCyclicGroup` as derived trait of FiniteGroup that defines a generator.

Reply:
You da bomb